torch/_inductor/graph.py

```diff
@@ -1236,6 +1213,53 @@ def significant_strides_equal(
         )
         return ir.TensorBox(torch._inductor.ir.ReinterpretView(storage, new_layout))
 
+    def propagate_mutation(
+        self,
+        fx_node: torch.fx.Node,
+        old_args: Tuple[Any],
+        old_kwargs: Dict[str, Any],
+        new_args: Tuple[Any],
+        new_kwargs: Dict[str, Any],
+    ) -> None:
+        assert isinstance(fx_node.target, torch._ops.OpOverload)
+        assert len(old_args) == len(new_args)
+        assert len(old_kwargs) == len(new_kwargs)
+
+        def maybe_propagate(
+            schema_arg: torch._C.Argument, old_arg: ir.IRNode, new_arg: ir.IRNode
+        ) -> None:
+            if old_arg is new_arg:
+                return
+            if schema_arg.alias_info is not None and schema_arg.alias_info.is_write:
+                # The lowering for copy_ is smart enough to "replace" old_arg with
+                # new_arg in all future uses so a copy_ kernel never gets emitted.
+                self.call_function(torch.ops.aten.copy_.default, (old_arg, new_arg), {})
+
+        schema = fx_node.target._schema
+        for idx, (old_arg, new_arg) in enumerate(zip(old_args, new_args)):
+            schema_arg = schema.arguments[idx]
+            maybe_propagate(schema_arg, old_arg, new_arg)
+
+        schema_kwargs = {arg.name: arg for arg in schema.arguments}
+
+        for key in old_kwargs.keys():
+            old_arg = old_kwargs[key]
+            new_arg = new_kwargs[key]
+            schema_arg = schema_kwargs[key]
+            maybe_propagate(schema_arg, old_arg, new_arg)
+
+    def maybe_add_custom_op_layout_constraints(self, op: Callable[..., Any]) -> None:
+        if not isinstance(op, torch._ops.OpOverload):
+            return
+        # needs_fixed_stride_order only applies to ops that are FallbackKernel.
+        # Those do not have lowerings pre-registered.
+        if op in lowerings:
+            return
+        # lowering an op that is FallbackKernel for the first time will automatically
+        # generate the lowering so that the following isn't called multiple times.
+        if torch._C.Tag.needs_fixed_stride_order in op.tags:
+            torch._inductor.lowering.add_layout_constraint(op, constrain_to_fx_strides)
```

Comment:
where is the logic that prevents this from being called twice ? tried to find and failed

Reply:
I'll make the comment clearer. The first time we call maybe_add_custom_op_layout_constraints with a custom op `foo`:
- there's no lowering for the custom op
- directly after this function, in run_node, we do GraphLowering.call_function
- This generates a lowering for the custom op (make_fallback in https://github.com/pytorch/pytorch/blob/2bffbe06bdf8f130159917c3ecda2c5852fd35b4/torch/_inductor/graph.py#L1023)

The second time we call maybe_add_custom_op_layout_constraints with a custom op `foo`: 
- there's a lowering for the custom op, so the early return prevents us from adding the layout constraint twice.

Note that this was how the code worked previously [(we would generate the lowering and add the layout constraint once on the first usage of the custom op)](https://github.com/pytorch/pytorch/blob/2bffbe06bdf8f130159917c3ecda2c5852fd35b4/torch/_inductor/graph.py#L1010-L1023) but I needed to hoist this logic out because resolving layout constraints happens in `run_node`

Reply:
I ended up adding new logic to register the layout constraint once to make it clearer what is going on. Relying on the lowering being populated seemed like too much mental overload